### ISSUE TITLE
JavaScript client code generator emits invalid code for map and array types

### ISSUE THREAD

**Issue by demonfiddler** [swagger-api/swagger-codegen#2471]
From YAML, a definition of `type: object` with `additionalProperties` or a  definition of `type: array` with `items: xxx` results in the emission of invalid code in the `constructFromObject()` method body:

```javascript
  exports.constructFromObject = function(data, obj) {
    if (data) { 
      obj = obj || new exports();
      // -- IN THE ARRAY CASE --
      Array&lt;ElementType&gt;.constructFromObject(data, obj); <== unparseable, syntax error
      // -- OR IN THE MAP CASE --
      Object&lt;String, ElementType&gt;.constructFromObject(data, obj); <== unparseable, syntax error
    }
    return obj;
  }
```
This happens because `DefaultCodegen.addParentContainer()` (called from `DefaultCodegen.fromModel(String, Model, Map<String, Model>)`) sets the `parent` of the `CodegenModel` (representing the YAML definition) to the instantiation type for the array element type in the array case or for the `additionalProperties` type in the map case. The templates emit invalid code based on misinterpretation of the `parent` setting.

At this stage it's unclear what the correct solution might be. I need to get this fixed promptly as our application depends upon both patterns, so I'll investigate then update this issue with a proposed solution and will later raise a PR.

UPDATE: Turns out it's straightforward enough to suppress the syntax error by guarding with an outer #parentModel section, but less easy to implement the correct logic for array and map types to reflect the intended array and map semantics. In the array case particularly, even ECMAScript 5.1 provides no good mechanism to extend the built-in Array class: Object.setPrototypeOf(new Array(),  Xxx) logs performance warning messages in Firefox, for example. The ES 6 `class Xxx extends Array` syntactic sugar' ought to work too, by some accounts.

**Comment by Nepoxx** [comment]
Is this still an issue?

**Comment by wing328** [comment]
@Nepoxx this issue is for JS.

Please open a new issue for tracking and provide the spec to repeat the issue.

**Comment by wing328** [comment]
Closed via #2445
